services/commands-service.ts

```diff
@@ -19,14 +19,27 @@ export class CommandsService implements ICommandsService {
 	private static HIERARCHICAL_COMMANDS_DEFAULT_COMMAND_DELIMITER = "|*";
 	private static HOOKS_COMMANDS_DELIMITER = "-";
 	private areDynamicSubcommandsRegistered = false;
+	private cachedCommandHelp: any = {};
 
-	constructor(private $errors: IErrors,
-		private $logger: ILogger,
-		private $injector: IInjector,
-		private $staticConfig: Config.IStaticConfig,
+	constructor(private $commandsServiceProvider: ICommandsServiceProvider,
+		private $errors: IErrors,
+		private $fs: IFileSystem,
 		private $hooksService: IHooksService,
-		private $commandsServiceProvider: ICommandsServiceProvider,
-		private $options: ICommonOptions) {
+		private $injector: IInjector,
+		private $logger: ILogger,
+		private $options: ICommonOptions,
+		private $resources: IResourceLoader,
+		private $staticConfig: Config.IStaticConfig) {
+	}
+
+	public getCommandHelp(): IFuture<any> {
+		return (() => {
+			if (_.isEmpty(this.cachedCommandHelp) && this.$fs.exists(this.$resources.resolvePath(this.$staticConfig.COMMAND_HELP_FILE_NAME)).wait()) {
```

Comment:
Actually this is a very common pattern - lazy instantiation. Consider creating a `@lazy` decorator in order to abstract this code.